### PR TITLE
Replace scmURL with scmURi and add schema for scm decorator

### PR DESCRIPTION
**Pull request by minz1027**
- Replace `scmUrl` with `scmUri` (e.g. github.com:123456:master) as key
- Add `checkoutUrl` which can be either `ssh` or `https`
- Add schema to validate the input and output for scm plugin methods

### CODE REVIEW

File: core/scm.js [hunk 1]
```diff
@@ -68,8 +57,54 @@ const SCHEMA_COMMIT = Joi.object().keys({
...
+    branch: Joi.string()
+        .required()
+        .label('Branch of the repository'),
+
+    prRef: Joi.string()
+        .optional()
+        .label('PR reference of the repository'),
+
+    sha: Joi.string().hex()
+        .length(40)
```
Comment (d2lam) [inline core/scm.js]: bitbucket does not always give back the full sha

File: core/scm.js [hunk 2]
```diff
@@ -68,8 +57,54 @@ const SCHEMA_COMMIT = Joi.object().keys({
...
+        .required()
+        .label('Commit SHA')
+        .example('ccc49349d3cffbd12ea9e3d41521480b4aa5de5f'),
+
+    username: Joi.string()
+        .required()
+        .label('Display Name')
```
Comment (d2lam) [inline core/scm.js]: This should be username, not Display Name. They are different.
d2lam vs Dao Lam

File: core/scm.js [hunk 3]
```diff
@@ -26,15 +27,15 @@ const SCHEMA_USER = Joi.object().keys({
 }).label('SCM User');
 
 const SCHEMA_REPO = Joi.object().keys({
-    id: Joi.string()
+    title: Joi.string()
```
Comment (d2lam) [inline core/scm.js]: I feel like the name `title` and `subtitle` is a little confusing. Any reason why we want to name them like that?

File: core/scm.js (hunk 1 above)
Comment (minz1027) [inline core/scm.js]: removed the length() now

File: core/scm.js (hunk 2 above)
Comment (minz1027) [inline core/scm.js]: Fixed

File: models/pipeline.js [hunk 4]
```diff
@@ -11,10 +11,16 @@ const MODEL = {
...
-        .example('git@github.com:screwdriver-cd/data-model.git#master'),
+    checkoutUrl: Joi
+        .string().regex(Regex.CHECKOUT_URL)
+        .description('Checkout url for the application')
+        .example('git@github.com:screwdriver-cd/data-schema.git#master')
+        .example('https://github.com/screwdriver-cd/data-schema.git#master'),
+
+    scmUri: Joi
+        .string().regex(Regex.SCM_URI)
+        .description('Unique idetifier for the application')
```
Comment (tkyi) [inline models/pipeline.js]: identifier*

**Pull request merged**